Add next-item query based on a join

bnaecker opened the pull request:
- Adds the `NextItemSelfJoined` query, which implements the next-item query as a self-join. The current implementation joins the target table with the series of all possible values for the target item. Because CRDB eagerly evaluates subqueries, this entire list must be buffered in memory. The self-join version instead joins the existing table with the literal next item (item + 1), and finds the lowest one that's not allocated yet. Some initial testing suggests this can consume much less memory and run much faster than the previous query implementations. As a tradeoff, this new query cannot be used to randomly select _any_ item in a range -- only the next, lowest value can be selected. Note that this changes the way we allocate MAC addresses, which were previously random.
- Fixes #5904

bnaecker commented:
In addition to the tests I've added here, I tried to get a better sense of the actual memory consumption of these two flavors of query. I added a skipped test that populates a big table, and then prints the queries that can be used to find the next item in either version. I then ran the queries via `EXPLAIN ANALYZE` manually (since CRDB doesn't support bind parameters in this case).

Here's the original, `generate_series()`-based implementation:

```
root@localhost:32221/omicron> EXPLAIN ANALYZE SELECT 0 + "shift" AS "value" FROM (SELECT generate_series(0, 10000) AS "index", generate_series(0, 10000) AS "shift") LEFT OUTER JOIN "test_schema"."item" ON ("value", "time_deleted" IS NULL)
                              info
----------------------------------------------------------------
  planning time: 3ms
  execution time: 833ms
  distribution: local
  vectorized: true
  rows read from KV: 20,000 (371 KiB)
  cumulative time spent in KV: 796ms
  maximum memory usage: 8.3 MiB
  network usage: 0 B (0 messages)

  • render
  │ nodes: n1
  │ actual row count: 1
  │ estimated max memory allocated: 10 KiB
  │ estimated max sql temp disk usage: 0 B
  │ estimated row count: 1
  │
  └── • top-k
      │ nodes: n1
      │ actual row count: 1
      │ estimated max memory allocated: 10 KiB
      │ estimated max sql temp disk usage: 0 B
      │ estimated row count: 1
      │ order: +generate_series
      │ k: 1
      │
      └── • filter
          │ nodes: n1
          │ actual row count: 1
          │ estimated row count: 1
          │ filter: value IS NULL
          │
          └── • lookup join (left outer)
              │ nodes: n1
              │ actual row count: 10,001
              │ KV time: 427ms
              │ KV contention time: 0µs
              │ KV rows read: 10,000
              │ KV bytes read: 58 KiB
              │ estimated max memory allocated: 3.2 MiB
              │ table: item@item_pkey
              │ equality: (id) = (id)
              │ equality cols are key
              │
              └── • lookup join (left outer)
                  │ nodes: n1
                  │ actual row count: 10,001
                  │ KV time: 369ms
                  │ KV contention time: 0µs
                  │ KV rows read: 10,000
                  │ KV bytes read: 313 KiB
                  │ estimated max memory allocated: 2.6 MiB
                  │ estimated row count: 10
                  │ table: item@item_value_key (partial index)
                  │ equality: (generate_series) = (value)
                  │
                  └── • project set
                      │ nodes: n1
                      │ actual row count: 10,001
                      │ estimated row count: 10
                      │
                      └── • emptyrow
                            nodes: n1
                            actual row count: 1
(63 rows)


Time: 837ms total (execution 837ms / network 1ms)

```

The main datapoint here is the memory usage, 8.3MiB. The other details are also interesting though. The query took nearly a second to run, which surprised me. Almost all of that is in the KV-layer. Since the plan here involves a lookup-join, I think that means it's basically spending all the time finding the actual tuple with each value from the output of `generate_series()`, using the partial index `item@item_value_key`.

Here's the join-based analysis:

```
root@localhost:32221/omicron> EXPLAIN ANALYZE SELECT IF(EXISTS(SELECT 1 FROM "test_schema"."item" WHERE "time_deleted" IS NULL LIMIT 1), (SELECT "next_item" AS "value" FROM (SELECT "value" + 1 AS "next_item" FROM "test_schema"."item" WHERE "time_deleted" IS NULL) LEFT OUTER JOIN "test_schema"."item" ON ("next_item", "next_item" <= 10000, "time_deleted" IS NULL) = ("value", TRUE, TRUE) WHERE "value" IS NULL AND "next_item" <= 10000 ORDER BY "next_item" LIMIT 1), 0);
                                                                                                                                                            info
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  planning time: 594µs
  execution time: 13ms
  distribution: local
  vectorized: true
  rows read from KV: 20,001 (977 KiB)
  cumulative time spent in KV: 9ms
  maximum memory usage: 1.2 MiB
  network usage: 0 B (0 messages)

  • root
  │
  ├── • values
  │     nodes: n1
  │     actual row count: 1
  │     size: 1 column, 1 row
  │
  ├── • subquery
  │   │ id: @S1
  │   │ original sql: EXISTS (SELECT 1 FROM test_schema.item WHERE time_deleted IS NULL LIMIT 1)
  │   │ exec mode: exists
  │   │
  │   └── • limit
  │       │ nodes: n1
  │       │ actual row count: 1
  │       │ KV time: 914µs
  │       │ KV contention time: 0µs
  │       │ KV rows read: 1
  │       │ KV bytes read: 48 B
  │       │ estimated max memory allocated: 20 KiB
  │       │ estimated row count: 1
  │       │ count: 1
  │       │
  │       └── • render
  │           │ nodes: n1
  │           │ actual row count: 1
  │           │ KV time: 914µs
  │           │ KV contention time: 0µs
  │           │ KV rows read: 1
  │           │ KV bytes read: 48 B
  │           │ estimated max memory allocated: 20 KiB
  │           │ estimated row count: 10,000
  │           │
  │           └── • scan
  │                 nodes: n1
  │                 actual row count: 1
  │                 KV time: 914µs
  │                 KV contention time: 0µs
  │                 KV rows read: 1
  │                 KV bytes read: 48 B
  │                 estimated max memory allocated: 20 KiB
  │                 estimated row count: 1 - 10,000 (100% of the table; stats collected 42 minutes ago)
  │                 table: item@item_value_key (partial index)
  │                 spans: FULL SCAN
  │
  └── • subquery
      │ id: @S2
      │ original sql: (SELECT next_item AS value FROM (SELECT value + 1 AS next_item FROM test_schema.item WHERE time_deleted IS NULL) LEFT JOIN test_schema.item ON (next_item, next_item <= 10000, time_deleted IS NULL) = (value, true, true) WHERE (value IS NULL) AND (next_item <= 10000) ORDER BY next_item LIMIT 1)
      │ exec mode: one row
      │
      └── • top-k
          │ nodes: n1
          │ actual row count: 1
          │ estimated max memory allocated: 10 KiB
          │ estimated max sql temp disk usage: 0 B
          │ estimated row count: 1
          │ order: +next_item
          │ k: 1
          │
          └── • filter
              │ nodes: n1
              │ actual row count: 1
              │ estimated row count: 1
              │ filter: value IS NULL
              │
              └── • hash join (right outer)
                  │ nodes: n1
                  │ actual row count: 10,000
                  │ estimated max memory allocated: 490 KiB
                  │ estimated max sql temp disk usage: 0 B
                  │ estimated row count: 3,360
                  │ equality: (value) = (next_item)
                  │
                  ├── • render
                  │   │ nodes: n1
                  │   │ actual row count: 10,000
                  │   │ KV time: 4ms
                  │   │ KV contention time: 0µs
                  │   │ KV rows read: 10,000
                  │   │ KV bytes read: 488 KiB
                  │   │ estimated max memory allocated: 510 KiB
                  │   │ estimated row count: 10,000
                  │   │
                  │   └── • scan
                  │         nodes: n1
                  │         actual row count: 10,000
                  │         KV time: 4ms
                  │         KV contention time: 0µs
                  │         KV rows read: 10,000
                  │         KV bytes read: 488 KiB
                  │         estimated max memory allocated: 510 KiB
                  │         estimated row count: 10,000 (100% of the table; stats collected 42 minutes ago)
                  │         table: item@item_value_key (partial index)
                  │         spans: [ - /10000]
                  │
                  └── • filter
                      │ nodes: n1
                      │ actual row count: 10,000
                      │ estimated row count: 3,333
                      │ filter: next_item <= 10000
                      │
                      └── • render
                          │ nodes: n1
                          │ actual row count: 10,000
                          │ estimated row count: 10,000
                          │
                          └── • render
                              │ nodes: n1
                              │ actual row count: 10,000
                              │ KV time: 5ms
                              │ KV contention time: 0µs
                              │ KV rows read: 10,000
                              │ KV bytes read: 488 KiB
                              │ estimated max memory allocated: 510 KiB
                              │ estimated row count: 10,000
                              │
                              └── • scan
                                    nodes: n1
                                    actual row count: 10,000
                                    KV time: 5ms
                                    KV contention time: 0µs
                                    KV rows read: 10,000
                                    KV bytes read: 488 KiB
                                    estimated max memory allocated: 510 KiB
                                    estimated row count: 10,000 (100% of the table; stats collected 42 minutes ago)
                                    table: item@item_value_key (partial index)
                                    spans: FULL SCAN
(136 rows)


Time: 17ms total (execution 14ms / network 3ms)

```

Again, the main point here is the memory usage, 1.2MiB. That's much lower than the series-based version. The execution time is also much faster, clocking in at 13ms. I'm not really going to try to interpret the query plan itself -- I just don't know enough about the innards to say anything meaningful about whether this new query will _always_ use less memory. It definitely does use less memory in this case.

smklein commented on nexus/db-queries/src/db/queries/next_item.rs:
```diff
@@ -584,13 +603,328 @@ impl<Item> ShiftGenerator<Item> for DefaultShiftGenerator<Item> {
+    /// Create a new `NextItemSelfJoined` query, scoped to a particular key.
+    pub(super) fn new_scoped(
+        scope_key: ScopeKey,
+        item_min: Item,
+        item_max: Item,
+    ) -> Self {
```
The intention is that these are both inclusive, right? Could we document that?

bnaecker replied:
Documented in 75311d5a7

smklein commented on nexus/db-queries/src/db/queries/next_item.rs:
```diff
@@ -584,13 +603,328 @@ impl<Item> ShiftGenerator<Item> for DefaultShiftGenerator<Item> {
+{
+    /// Create a new `NextItemSelfJoined` query, with a global scope.
+    //
+    // NOTE: This is currently used only in tests.
```
Do we ever need this outside the crate? `#[cfg(test)]` should let us omit the `dead_code` warning suppression

bnaecker replied:
We don't need it out of this crate, but we do use the unscoped version of the original query today. I'll change the lint though, that's better.

bnaecker replied:
Changed to `cfg(test)` in 75311d5a7

smklein commented on nexus/db-queries/src/db/queries/next_item.rs:
```diff
@@ -584,13 +603,328 @@ impl<Item> ShiftGenerator<Item> for DefaultShiftGenerator<Item> {
+///             1
+///         FROM
+///             <table>
+///         WHERE
+///             time_deleted IS NULL
```
Aren't we also checking the scope here?

In the code below, I see: `WHERE ScopeColumn::NAME = scope_key`

bnaecker replied:
Yes, added that in 75311d5a7

smklein commented on nexus/db-queries/src/db/queries/next_item.rs:
```diff
@@ -700,6 +1034,68 @@ mod tests {
+
+    // These implementations are needed to actually allow inserting the results
+    // of the `NextItemSelfJoinedQuery` itself.
+    impl NextItemSelfJoinedQuery {
```
I know this is test-only code, but could we supply the bounds to `new()` as arguments? Took me a moment to realize this was bound to the range `[0, 9]`

bnaecker replied:
Added arguments in 75311d5a7

smklein commented on nexus/db-queries/src/db/queries/next_item.rs:
```diff
@@ -584,13 +603,328 @@ impl<Item> ShiftGenerator<Item> for DefaultShiftGenerator<Item> {
+///             (<scope_key>, TRUE, TRUE)
+///         WHERE
+///             <item_column> IS NULL AND next_item <= <max_item>
+///         ORDER BY next_item
```
Does the `ORDER BY next_item` matter here? We can keep it in for tests determinism, but it basically means "we're picking the lowest possible free item, immediately after all existing items", right?

It's not randomly distributed, but omitting this would mean "we could pick *any* next lowest possible free item after all existing items", right?

bnaecker replied:
The ordering was intentional, yeah. It doesn't quite mean "after all existing items", it really means "the lowest possible free item". Those differ if there are gaps, for example. We fill the gaps first, with this query.

I think that's usually what we want to do. The only place we don't already do that is when we allocate a VNI for a new VPC (which this PR doesn't change) and when we select a MAC for a NIC (which this PR _does_ change). I don't have a strong opinion about whether it _matters_, but I do like the determinism for most situations.

bnaecker commented:
Thanks for the eyes @smklein! I've addressed your comments, except for the point about ordering. I personally like the determinism, but I admit it changes how MAC addresses for NICs are currently allocated (they were random before). I don't think that's load-bearing, and I think the determinism is useful for many other situations (IP addresses), so I'd prefer to keep it for now unless you strongly disagree.

smklein commented:
> Thanks for the eyes @smklein! I've addressed your comments, except for the point about ordering. I personally like the determinism, but I admit it changes how MAC addresses for NICs are currently allocated (they were random before). I don't think that's load-bearing, and I think the determinism is useful for many other situations (IP addresses), so I'd prefer to keep it for now unless you strongly disagree.

Totally fine keeping as-is -- if we want random allocation, I think we'll need to build that fully intentionally, and I understand that having determinisim helps with tests.
